### ISSUE TITLE
Make it possible to catch specific error types and chain error handling

### ISSUE THREAD

**Issue by gwynne** [vapor/async#65]
Example:
```swift
func getUser(req: Request) -> Future<ResponseRepresentable> {
    return User.find(req.parameters.next()).map(to: PublicUser.self) { user in
        return try user.makePublic()
    }.catchMap(UserError.self, to: Response.self) { error in
        guard case .makePublicFailed = error else { throw error } 
        
        return req.makeResponse().withEncodedContent(MyErrorResponse())
    }
}
```
Of particular interest is how to handle `catchMap()` blocks returning a different type than the last `map()` block.

**Comment by calebkleveter** [comment]
Why call it `catchMap`? If I am correct, we want to mirror `do/catch` syntax so it feels more natural. I am more in favor of changing (or overloading) the current `catch` method to handle throwing and returning.

**Comment by jdmcd** [comment]
I actually prefer `catchMap`. It still returns a value (in @gwynne's example, `return req.makeResponse().withEncodedContent(MyErrorResponse())`) which I think is reason enough to let it follow the format of the current vapor `map` and `flatMap` methods. _Especially_ if it's going to have similar overloads like `to: `. 

Regular `catch` might get confusing with Vapor's `catch`, as the Swift `catch` doesn't explicitly return a value.

**Comment by calebkleveter** [comment]
Correct, it doesn't return a value, but I can return a value from a method from inside the `catch` block (or throw, for that matter).

**Comment by jdmcd** [comment]
Of course, but the method proposed here _requires_ some kind of return, since it's chained on to the map or flatMap above it. Personally I think it's clearer when it follows the same pattern

**Comment by kdawgwilk** [comment]
I would prefer to keep things as similar to RxSwift as we can so the transition is easy for iOS devs using RxSwift to come over to Vapor 3: `catchError` && `catchErrorJustReturn`
